citation-categories/java/src/main/java/org/linqs/psl/examples/citationcategories/Run.java

```diff
@@ -0,0 +1,252 @@
+        Partition targetsPartition = dataStore.getPartition(PARTITION_LEARN_TARGETS);
+        Partition truthPartition = dataStore.getPartition(PARTITION_LEARN_TRUTH);
+
+        // This database contains all the ground atoms (targets) that we want to infer.
+        // It also includes the observed data (because we will run inference over this db).
+        Set<StandardPredicate>set = new HashSet<StandardPredicate>();
+        set.add(model.getStandardPredicate("Link"));
+        Database randomVariableDatabase = dataStore.getDatabase(targetsPartition, set, obsPartition);
+
+        // This database only contains the true ground atoms.
+        Database observedTruthDatabase = dataStore.getDatabase(truthPartition, dataStore.getRegisteredPredicates());
+
+        WeightLearningApplication wla = new MaxLikelihoodMPE(model, randomVariableDatabase, observedTruthDatabase);
+        wla.learn();
+
+        randomVariableDatabase.close();
+        observedTruthDatabase.close();
+
+        log.info("Weight learning complete");
+    }
+
+    /**
+     * Run inference to infer the unknown HasCat relationships between people.
+     */
+    private void runInference() {
+        log.info("Starting inference");
+
+        Partition obsPartition = dataStore.getPartition(PARTITION_EVAL_OBSERVATIONS);
+        Partition targetsPartition = dataStore.getPartition(PARTITION_EVAL_TARGETS);
+        Set<StandardPredicate>set = new HashSet<StandardPredicate>();
+        set.add(model.getStandardPredicate("Link"));
+        Database inferDB = dataStore.getDatabase(targetsPartition, set, obsPartition);
+
+        InferenceApplication inference = new MPEInference(model, inferDB);
+        inference.inference();
+        inference.close();
+        inferDB.close();
+
+        log.info("Inference complete");
+    }
+
+    /**
+     * Run statistical evaluation scripts to determine the quality of the inferences
+     * relative to the defined truth.
+     * Note that the target predicate is categorical and we will assign the category with the
+     * highest truth value as true and the rest false.
+     */
+    private void evalResults() {
+        // Because the truth data also includes observed data, we will make sure to include the observed
+        // partition here.
+        Set<StandardPredicate>set = new HashSet<StandardPredicate>();
+        set.add(model.getStandardPredicate("Link"));
+        Database resultsDB = dataStore.getDatabase(dataStore.getPartition(PARTITION_EVAL_TARGETS), set, dataStore.getPartition(PARTITION_EVAL_OBSERVATIONS));
+        Database truthDB = dataStore.getDatabase(dataStore.getPartition(PARTITION_EVAL_TRUTH), dataStore.getRegisteredPredicates());
+
+        Evaluator eval = new CategoricalEvaluator(1);
+        eval.compute(resultsDB, truthDB, model.getStandardPredicate("HasCat"));
+        log.info(eval.getAllStats());
+
+        resultsDB.close();
+        truthDB.close();
+    }
+    
```

Comment:
There is extra whitespace at the end of this line (this occurs in other files you submitted as well).
```
find . -name Run.java | xargs grep -nE '\s+$'
```